### DISCUSSION TITLE
Pooling of Streams in Marten.AspNetCore.QueryableExtensions

### DISCUSSION THREAD

**Discussion by gfoidl** [JasperFx/marten#2408] [Ideas]
For each request a [`new MemoryStream`](https://github.com/JasperFx/marten/blob/217cdbc835ad1086375db9a041f13c3f8821c2a6/src/Marten.AspNetCore/QueryableExtensions.cs#L23) is created, thus also the backing `byte[]`-buffer used by MS and if it needs to be resized another buffer with copying over the old contents, and so on.

As response header values can't be set once the body has started, any kind of temporary stream needs to be used -- as it's done now.
But instead of creating a MS every time there should be some kind of pooling, e.g. via [Microsoft.IO.RecyclableMemoryStream](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream).

Is this in general something to look into, and is it OK to rely on that package?
Otherwise there could be some inbox-solution for pooled streams based on ArrayPool, etc.

I didn't run any benchmarks so far, just saw this when reading the code.

<sub>I can create a PR if needed.</sub>.

**Comment by cocowalla** [comment]
I use that package in several projects, it's been rock solid. If for some reason it wasn't fit for Marten, it would be trivial to build something on top of [Microsoft.Extensions.ObjectPool](https://www.nuget.org/packages/Microsoft.Extensions.ObjectPool) instead.

**Reply by gfoidl** [reply]
With MemoryStream and the ObjectPool one has to be cautious about LOH and all other "troubles" pooling brings with it.
So the easiest would be to use recycleable MS linked above IMO.

**Comment by gfoidl** [comment]
Also where possible one could omit to set the _ContentLength_, e.g.:
https://github.com/JasperFx/marten/blob/b77d5537327d5e69710cceafc90bd2fb43c3e49f/src/Marten.AspNetCore/QueryableExtensions.cs#L49-L61

ASP.NET Core will produce a "chunked" response then, and afaik every client can handle that.
The advantage is that `StreamJsonArray` could write to the response-stream directly, thus avoiding the need of the MemoryStream (even when pooled) and the copying of MS to response.
